File: phone/src/main/java/com/problem/phone/Phone.java
package com.problem.phone;

import static java.lang.Character.isDigit;

public class Phone {
	public static String solution(String input) {
		StringBuilder sb = input
			.chars()
			.filter(c -> isDigit(c))
			.collect(
				() -> new StringBuilder(),
				(acc, i) -> {
					int L = acc.length();
					if (L > 3 && L % 4 == 0) acc.insert(L - 1, '-');
					acc.append((char) i);
				},
				null);

		int L = sb.length();
		if (L > 3 && L % 4 == 0) sb.insert(L - 2, '-');

        return sb.toString();
    }
}

File: phone-scala/src/main/scala/com/problem/phone/Phone.scala
package com.problem.phone

object Phone {
  def process(input: String): String = {
    val res = new StringBuilder()

    input
      .filter(Character.isDigit)
      .grouped(3)
      .foreach { group =>
        if (!res.isEmpty) {
          if (group.length > 1)
            res += '-'
          else // corner case for two 2-sized blocks
            res.insert(res.length - 1, '-')
        }

        res ++= group
      }

    res.toString
  }
}
